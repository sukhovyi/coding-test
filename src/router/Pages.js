const Pages = Object.freeze({
  MANAGE_POSTS: 'MANAGE_POSTS',
  CREATE_POST: 'CREATE_POST',
  EDIT_POST: 'EDIT_POST',
});

export default Pages;
