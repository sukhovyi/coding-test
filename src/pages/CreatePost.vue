<template>
  <ui-row>
    <ui-col>
      <h1>
        Create Post
      </h1>
    </ui-col>
    <ui-col>
      <post-form
        :post="post"
        @save="createPost"
        @cancel="cancel"
      />
    </ui-col>
  </ui-row>
</template>

<script>
import PostForm from '../components/PostForm';
import UiButton from '../ui-kit/ui-button/UiButton';
import Pages from '../router/Pages';
import UiInput from '../ui-kit/ui-input/UiInput';
import { CREATE_POST } from '../apollo/mutations';
import UiRow from '../ui-kit/ui-row/UiRow';
import UiCol from '../ui-kit/ui-col/UiCol';

export default {
  name: 'CreatePost',
  components: { UiCol, UiRow, UiInput, UiButton, PostForm },
  data() {
    return {
      post: {
        title: null,
        body: null,
      },
    };
  },
  methods: {
    async createPost(post) {
      await this.$apollo.mutate({
        mutation: CREATE_POST,
        variables: {
          input: {
            title: post.title,
            body: post.body,
          },
        },
      });
      this.returnBack();
    },
    cancel() {
      this.returnBack();
    },
    returnBack() {
      this.$router.push({ name: Pages.MANAGE_POSTS });
    },
  },
};
</script>
