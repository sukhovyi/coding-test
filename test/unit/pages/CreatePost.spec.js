import Vue from 'vue';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import CreatePost from '../../../src/pages/CreatePost';
import Pages from '../../../src/router/Pages';

jest.mock('../../../src/apollo/mutations', () => ({
  CREATE_POST: 'CREATE_POST',
}));

describe('pages/CreatePost', () => {
  const localVue = createLocalVue();
  let mutate;
  let success;
  let push;
  let wrapper;

  beforeEach(() => {
    mutate = jest.fn();
    success = jest.fn();
    push = jest.fn();
    wrapper = shallowMount(CreatePost, {
      localVue,
      mocks: {
        $apollo: {
          mutate,
        },
        $toasted: {
          success,
        },
        $router: {
          push,
        },
      },
    });
  });


  it('should be rendered correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('createPost should works correctly', async () => {
    const post = {
      title: 'title',
      body: 'body',
    };
    wrapper.vm.createPost(post);
    await Vue.nextTick();

    expect(mutate).toBeCalledWith({ mutation: 'CREATE_POST', variables: { input: { ...post } } });
    expect(success).toBeCalledWith('Post was successfully created.');
    expect(push).toBeCalledWith({ name: Pages.MANAGE_POSTS });
  });

  it('cancel should works correctly', async () => {
    wrapper.vm.cancel();

    expect(push).toBeCalledWith({ name: Pages.MANAGE_POSTS });
  });

  it('returnBack should works correctly', async () => {
    wrapper.vm.returnBack();

    expect(push).toBeCalledWith({ name: Pages.MANAGE_POSTS });
  });
});
