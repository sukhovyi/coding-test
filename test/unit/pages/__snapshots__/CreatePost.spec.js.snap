// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/CreatePost should be rendered correctly 1`] = `
<ui-row-stub>
  <ui-col-stub
    cols="10"
  >
    <h1>
      
      Create Post
    
    </h1>
  </ui-col-stub>
   
  <ui-col-stub
    cols="10"
  >
    <post-form-stub
      post="[object Object]"
    />
  </ui-col-stub>
</ui-row-stub>
`;
